query_execution/QueryContext.cpp

```diff
@@ -58,7 +59,8 @@ QueryContext::QueryContext(const serialization::QueryContext &proto,
                            const CatalogDatabaseLite &database,
                            StorageManager *storage_manager,
                            const tmb::client_id scheduler_client_id,
-                           tmb::MessageBus *bus) {
+                           tmb::MessageBus *bus)
+    : database_(static_cast<const CatalogDatabase&>(database)) {
```

Comment:
Again, DataPipelineMessage is only generated during pipelining or when a block gets full. Thus it doesn't cover all the cases. I am not sure if you have any fundamental issue with this design. It's actually pretty straightforward.